PR - Test 121 login sign up

darkheros12 opened the pull request:
### Related Issues
#121 

### Description
I added espresso test case for login and sign up.
It will test directly to the database as it has to go through multiple activities.
I used dumb email and random phone number.

### How to check PR
please check file LoginPageActivityTest and SignUpPageActivityTest

darkheros12 committed:
first initial commit

set up login page testing

darkheros12 committed:
Added Login and SignUp Test

- Added Espresso Test cases for login and sign up page

darkheros12 committed:
Removed unnecessary code

I removed unnecessary  from login page and sign up page activities because the test does not required to settings.

refatK commented on app/src/androidTest/java/com/example/bigowlapp/activity/SignUpPageActivityTest.java:
```diff
@@ -0,0 +1,140 @@
+import static org.mockito.Mockito.verify;
+
+@RunWith(AndroidJUnit4.class)
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
```
Just curious what this does.

darkheros12 replied:
This method test the test case alphabetically. But I think what you mention above, the CI instance are clear. My test case test directly the database that's why the test D did not passed. As I am required to test the database to make sure the login component works.

refatK commented on app/src/androidTest/java/com/example/bigowlapp/activity/SignUpPageActivityTest.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    @Test
+    public void A_unSufficientSignUpInputTest() {
```
I personally don't like this format of the method name, I think the descriptive names are enough and the letters at the begiinning don't seem to add any value. You can argue why you think it's valid though. Also starts with capital letter which isn't java standared for methods.

darkheros12 replied:
I know what you mean, but as I mentioned above, it for the test be test in alphabetical order.

darkheros12 replied:
I removed it.

refatK reviewed (changes requested):
You'll need to address the failing tests pointed out here, but overall they seem good:
https://github.com/Landry333/Big-Owl/pull/263/checks?check_run_id=2179839300#step:8:218

Also, if you want to be sure the test will work on the CI, make sure to clear the app's storage on the test device first, and then run your tests. By default, the CI will start with a fresh emulator and a fresh instance of the app. Sometimes we've come across issues where we were already logged in to the app, but since the CI can't do that, the tests end up failing when run on CI.

![image](https://user-images.githubusercontent.com/31963426/112402001-15225080-8ce2-11eb-9782-5ff951183c23.png)

darkheros12 committed:
fix PR commit

darkheros12 committed:
fix PR commit

darkheros12 committed:
fix PR commit

Change all the replace text to type text to simulate the real world situation

darkheros12 commented:
Okay, I found out why my sign up page test is not working. The app is not SCROLLABLE. I tested it with the smallest device QVGA 240 x 320. Therefore, those who have small devices cannot sign up.

darkheros12 committed:
Fix login test

refatK commented:
> Okay, I found out why my sign up page test is not working. The app is not SCROLLABLE. I tested it with the smallest device QVGA 240 x 320. Therefore, those who have small devices cannot sign up.

@darkheros12 Oh, if you think it's important, you can change the xml to make that page scrollable. Good catch though. I think a user during the usability test wanted it to be scrollable anyways, so would be a good change.

refatK commented:
@darkheros12 Pretty close, now just one failing

https://github.com/Landry333/Big-Owl/pull/263/checks?check_run_id=2196706387#step:8:233

![image](https://user-images.githubusercontent.com/31963426/112548228-435f6900-8d92-11eb-9c98-c2a1a2f577e0.png)

darkheros12 commented:
> @darkheros12 Pretty close, now just one failing
> 
> https://github.com/Landry333/Big-Owl/pull/263/checks?check_run_id=2196706387#step:8:233
> 
> ![image](https://user-images.githubusercontent.com/31963426/112548228-435f6900-8d92-11eb-9c98-c2a1a2f577e0.png)

that is because I commented them. I forgot to remove the comment.

darkheros12 committed:
Merge branch 'staging' into TEST-121-Login-SignUp

darkheros12 committed:
Fix PR test case and sign up layout

I added a scrollview on the sign up layout for smaller devices.

all the test should pass now.

refatK commented:
Will check locally, but nice job, 4.35% increase to coverage!

ShawnGuerra-Bautista commented on app/src/androidTest/java/com/example/bigowlapp/activity/SignUpPageActivityTest.java:
```diff
@@ -0,0 +1,202 @@
+
+    SignUpPageActivity signUpPageActivity;
+
+    public void setUp() {
```
Should have @Before annotation on this method.

refatK replied:
In case your having issues, it's `@Before` to be specific

darkheros12 replied:
it is added

ShawnGuerra-Bautista commented on app/src/androidTest/java/com/example/bigowlapp/activity/SignUpPageActivityTest.java:
```diff
@@ -0,0 +1,202 @@
+
+@RunWith(AndroidJUnit4.class)
+@LargeTest
+public class SignUpPageActivityTest {
```
The current coverage is about 63% for this class. Looking at the code coverage of the class, I think you can still add more tests that require little effort which can increase coverage even further.

![image](https://user-images.githubusercontent.com/23643192/112911358-4379a480-90c3-11eb-8f26-0259554e7238.png)

The one that I'm talking about is the error handling. It should be simple and it would cover lots of lines.

You would have code like the one in ScheduleFormFragmentTest: 
![image](https://user-images.githubusercontent.com/23643192/112910975-74a5a500-90c2-11eb-82a5-13bba9d734b8.png)

I would also add tests that would simply click the tvSignIn button.

Another thing, if it's possible, then I think mocking the viewModel would be beneficial.

darkheros12 replied:
I did an extra test for tvSignIn and I made a mocking now.

ShawnGuerra-Bautista commented on app/src/androidTest/java/com/example/bigowlapp/activity/LoginPageActivityTest.java:
```diff
@@ -0,0 +1,117 @@
+
+@RunWith(AndroidJUnit4.class)
+@LargeTest
+public class LoginPageActivityTest {
```
I believe the coverage is fine here.

Although, I think you can get more by mocking the viewModel. However, if it proves too difficult, then I think it's still good.

darkheros12 replied:
I made the required change

ShawnGuerra-Bautista reviewed (commented):
Very good. I like that the coverage has increase by about 5%.

There are some changes I request though.

ShawnGuerra-Bautista reviewed (changes requested):
I added my comments for changes above.

refatK commented on app/src/androidTest/java/com/example/bigowlapp/activity/SignUpPageActivityTest.java:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    @Test
+    public void validSignUpInputTest() {
+        //create randomness for phone number, test purpose only.
+        Random rand = new Random();
+        int min = 0;
+        int max = Integer.MAX_VALUE;
+        int result = rand.nextInt(max - min) + min;
+
+        this.phone = String.valueOf(result);
+        this.email = result + "@email.com";
+
+        onView(withId(R.id.user_first_name)).perform(typeText(firstName), ViewActions.closeSoftKeyboard());
+        onView(withId(R.id.user_last_name)).perform(typeText(lastName), ViewActions.closeSoftKeyboard());
+        onView(withId(R.id.edit_text_text_mail_address)).perform(typeText(this.email), ViewActions.closeSoftKeyboard());
+        onView(withId(R.id.edit_text_text_password)).perform(typeText(password), ViewActions.closeSoftKeyboard());
+        onView(withId(R.id.edit_text_phone)).perform(typeText(this.phone), ViewActions.closeSoftKeyboard());
+
+        onView(withId(R.id.button_sign_up))
+                .perform(scrollTo())
+                .check(matches(withText("Sign Up"))).perform(click());
+    }
```
This is a good test, but the issue is it pops in dummy users to the real database. SignUpViewModel should be mocked to prevent this from happening, and upon doing that, I verify can be added to confirm the creation of user is actually completed.

You can check out some of the other tests to see how that done, since we have similar ideas through out where the ViewModel of an activity/fragment is mocked, and instead of actually allowing it to touch the db, we instead verify calls onto the view models.

refatK commented on app/src/androidTest/java/com/example/bigowlapp/activity/LoginPageActivityTest.java:
```diff
@@ -0,0 +1,117 @@
+                .check(matches(withText("Sign In"))).perform(click());
+    }
+
+    private ViewAction waitFor(final long ms) {
+        return new ViewAction() {
+            @Override
+            public Matcher<View> getConstraints() {
+                return ViewMatchers.isRoot();
+            }
+
+            @Override
+            public String getDescription() {
+                return "Wait";
+            }
+
+            @Override
+            public void perform(UiController uiController, View view) {
+                SystemClock.sleep(ms);
+            }
+        };
+    }
```
This is unused so you can remove it

refatK reviewed (changes requested):
Once again, good job on getting the coverage up. My only issue is with one of the sign up tests adding users to the real database, but this is pretty easy to fix. After that I will approve.

darkheros12 committed:
Fix PR comments

- added mock view for login and sign up
- added set for activities

darkheros12 committed:
Update SignUpPageActivityTest.java

change Mock createUser to any()

refatK reviewed (approved):
All tests are passing, and you did my requests so approved!

ShawnGuerra-Bautista reviewed (approved):
Perfect. It has good coverage and everything is working fine.
